keras_nlp/tokenizers/byte_pair_tokenizer.py

```diff
@@ -0,0 +1,460 @@
+
+        The end condition is either the word has been fully merged (list has
+        only one byte string), or it can no longer perform a merge.
+        """
+        # Get all word pairs.
+        first, second = words[:, :-1], words[:, 1:]
+
+        # Mask empty.
+        non_empty_mask = second.nested_row_lengths()[0] != 0
+        mask = tf.logical_and(mask, non_empty_mask)
+        if tf.math.count_nonzero(mask) == 0:
+            return [words, mask]
+        non_empty_idxs = tf.boolean_mask(tf.range(tf.shape(mask)[0]), mask)
+        tmp_first = tf.ragged.boolean_mask(first, mask)
+        tmp_second = tf.ragged.boolean_mask(second, mask)
+
+        # Get byte pair ranking in merge rules.
+        pairs = tf.strings.join([tmp_first, tmp_second], separator=" ")
+        pair_rank = self.bpe_ranks.lookup(pairs)
+
+        # Get BPE pair ranks.
+        min_pair_rank = tf.reduce_min(pair_rank, axis=1)
+        not_found_mask = min_pair_rank != self.max_bpe_rank
+
+        # Tokens cannot be further merged are marked as finished.
+        mask = tf.tensor_scatter_nd_update(
+            mask, tf.expand_dims(non_empty_idxs, axis=1), not_found_mask
+        )
+        if tf.math.count_nonzero(mask) == 0:
+            return [words, mask]
+
+        masked_pair_rank = tf.ragged.boolean_mask(pair_rank, not_found_mask)
+        min_pair_rank_idx = tf.math.argmin(
+            masked_pair_rank.to_tensor(self.max_bpe_rank), axis=1
+        )
+
+        # Get words and pairs to process.
+        unfinished_words = tf.ragged.boolean_mask(words, mask)
+
+        pair_left = tf.gather(unfinished_words, min_pair_rank_idx, batch_dims=1)
+        pair_right = tf.gather(
+            unfinished_words, min_pair_rank_idx + 1, batch_dims=1
+        )
+
+        merged_pairs = tf.strings.join([pair_left, pair_right])
+        empty_strs = tf.fill(tf.shape(merged_pairs), "")
+
+        unfinished_word_indices = tf.cast(
+            tf.boolean_mask(tf.range(tf.shape(mask)[0]), mask), dtype=tf.int64
+        )
+        merged_pair_indices = tf.concat(
+            [
+                unfinished_word_indices[:, tf.newaxis],
+                min_pair_rank_idx[:, tf.newaxis],
+            ],
+            axis=1,
+        )
+        empty_string_indices = tf.concat(
+            [
+                unfinished_word_indices[:, tf.newaxis],
+                min_pair_rank_idx[:, tf.newaxis] + 1,
+            ],
+            axis=1,
+        )
+
+        tensor_words = words.to_tensor(default_value="")
+        tensor_words = tf.tensor_scatter_nd_update(
+            tensor_words,
+            merged_pair_indices,
+            merged_pairs,
+        )
+
+        words = tf.tensor_scatter_nd_update(
+            tensor_words,
+            empty_string_indices,
+            empty_strs,
+        )
+        words = self._remove_empty_strings(words)
+        return [words, mask]
+
+    def _byte_pair_encoding(self, tokens):
```

Comment:
I think, readability wise, we might want to make a separate `byte_pair_tokenizer_utils.py` file, and make a standalone function that is called `byte_pair_encode(tokens, bpe_ranks)`. Then we could pull this function, `_byte_pair_merge_loop_body`, `_encode_tokens`, `_remove_empty_strings` and `_remove_strings` all into that util file.

That would keep two main pieces of complexity separate. This file handles the tokenizer implementation, the regex splitting, and the cache. The util file handles the BPE algorithm.

What do you think?